_posts/2018-04-15-customreaders.md

```diff
@@ -0,0 +1,223 @@
+
+
+```python
+import MDAnalysis as mda
+
+u = mda.Universe('myfile.xyzc', format='XYZ')
+```
+
+
+    ---------------------------------------------------------------------------
+
+    EOFError                                  Traceback (most recent call last)
+
+    <ipython-input-2-19807941e521> in <module>()
+          1 import MDAnalysis as mda
+          2 
+    ----> 3 u = mda.Universe('myfile.xyzc', format='XYZ')
+    
+
+    /home/richard/miniconda2/envs/mydev/lib/python2.7/site-packages/MDAnalysis/core/universe.pyc in __init__(self, *args, **kwargs)
+        268             else:
+        269                 coordinatefile = args[1:]
+    --> 270             self.load_new(coordinatefile, **kwargs)
+        271 
+        272         # Check for guess_bonds
+
+
+    /home/richard/miniconda2/envs/mydev/lib/python2.7/site-packages/MDAnalysis/core/universe.pyc in load_new(self, filename, format, in_memory, **kwargs)
+        416         kwargs['n_atoms'] = self.atoms.n_atoms
+        417 
+    --> 418         self.trajectory = reader(filename, **kwargs)
+        419         if self.trajectory.n_atoms != len(self.atoms):
+        420             raise ValueError("The topology and {form} trajectory files don't"
+
+
+    /home/richard/miniconda2/envs/mydev/lib/python2.7/site-packages/MDAnalysis/coordinates/XYZ.pyc in __init__(self, filename, **kwargs)
+        309         # (Also cannot just use seek() or reset() because that would break
+        310         # with urllib2.urlopen() streams)
+    --> 311         self._read_next_timestep()
+        312 
+        313     @property
+
+
+    /home/richard/miniconda2/envs/mydev/lib/python2.7/site-packages/MDAnalysis/coordinates/XYZ.pyc in _read_next_timestep(self, ts)
+        369             return ts
+        370         except (ValueError, IndexError) as err:
+    --> 371             raise EOFError(err)
+        372 
+        373     def _reopen(self):
+
+
+    EOFError: could not convert string to float: red
+
+
+Here, as expected, the XYZReader chokes on the word 'red' where it expects a floating point number.
+
+We can rewrite the XYZReader to anticipate the extra column:
+
+
+```python
+import numpy as np
+from MDAnalysis.coordinates.XYZ import XYZReader
+from MDAnalysis.topology.XYZParser import XYZParser
+
+class XYZCReader(XYZReader):
+    format = 'XYZC'
+    
+    def _read_next_timestep(self, ts=None):
+        # check that the timestep object exists
+        if ts is None:
+            ts = self.ts
+
+        f = self.xyzfile
+
+        try:
+            # we assume that there are only two header lines per frame
+            f.readline()
+            f.readline()
+            for i in range(self.n_atoms):
+                self.ts._pos[i] = np.float32(f.readline().split()[2:5])
+            ts.frame += 1
+            return ts
+        except (ValueError, IndexError) as err:
+            raise EOFError(err)
+            
+class XYZCParser(XYZParser):
```

Comment:
I wouldn't have this here and yet. You haven't talked about the topology parsing yet. I think it's better to show an error and say that we can either give a topology file mdanalysis understands or this short class to add a xyzc topology format.